# Napier's Logarithms: 九九のない世界

* [John Napier - Wikipedia](https://en.wikipedia.org/wiki/John_Napier)
  * [Napierian logarithm - Wikipedia](https://en.wikipedia.org/wiki/Napierian_logarithm#:~:text=The%20term%20Napierian%20logarithm%20or,it%20is%20named%20after%20him.)
  * [Logarithms: The Early History of a Familiar Function - John Napier Introduces Logarithms | Mathematical Association of America](https://www.maa.org/press/periodicals/convergence/logarithms-the-early-history-of-a-familiar-function-john-napier-introduces-logarithms)

Napierは、$N$に対して対数 (logarithm) $L$を次の数式で定義し、$R=7$ の場合の数表を作成した。

$$
  \frac{N}{10^R} = \left(1-\frac{1}{10^{R}}\right)^{L}
$$

* [e (mathematical constant) - Wikipedia](https://en.wikipedia.org/wiki/E_%28mathematical_constant%29)

その後、Eulerは自然対数$e$を次の極限式で定義した。

$$
  \lim_{n\rightarrow\infty} \left( 1+\frac{1}{n} \right)^n = e
$$

$$
  \lim_{n\rightarrow\infty} \left( 1-\frac{1}{n} \right)^n = \frac{1}{e}
$$

$R$が十分大きいとき次の近似式が得られる。

$$
\left( 1-\frac{1}{10^R} \right)^{10^R} = (1-10^{-R})^{10^R} \approx \frac{1}{e}
$$

In [4]:
import math
(1-10**-5)**(10**5), (1-10**-7)**(10**7), 1/math.e

(0.3678776017682465, 0.367879422971105, 0.36787944117144233)

現代数学の$\log$を用いるとNapierの対数$L$は次のように近似できる。

$$
  L = \log_{1-10^{-R}} \frac{N}{10^R} \approx
  10^R \log_{1/e}\left( \frac{N}{10^R} \right) \approx -10^R \log_{e}\left( \frac{N}{10^R} \right)
$$

## Napierの定義

In [7]:
1-10**-5, 1-10**-7

(0.99999, 0.9999999)

$p_{n}$ を次の数列として定義したとき、

$$
\begin{eqnarray}
p_{0}   &=& 10^7 = 10000000 \\
p_{1}   &=& 10^7 \left(1-\frac{1}{10^7} \right)= 10000000 \times 0.9999999 = 9999999 \\
p_{n+1} &=& p_{n} \left(1-\frac{1}{10^7}\right) = p_{n} \times 0.9999999 
\end{eqnarray}
$$

数列の添字 $n$ は $p_{n}$ のNapier対数である。

$$
p_{n} = 10^7 \times \left(1-\frac{1}{10^7}\right)^n
$$

$$
  n = \log_{1-10^{-7}}\left(\frac{p_{n}}{10^7}\right)
$$

次のように対数法則を満たす。

$$
p_{m+n} = 10^R \times \left(1-\frac{1}{10^7}\right)^{m+n} = 10^R \times \left(1-\frac{1}{10^7}\right)^m \times \left(1-\frac{1}{10^7}\right)^n = p_{m} \times p_{n} \times \frac{1}{10^R}
$$

### $R=5$のとき

In [61]:
import math

def napier5(n):
    """
    R=5のときのNapierの数表の第n項を計算する
    """
    return 10**5 * (1-1/(10**5))**n

In [76]:
# `math.log()`を使って、Napierの対数を計算する
list(map(lambda x: math.log(x/10**5, 1-1/(10**5)), (1234, 5678)))

[439488.72859419696, 286855.6785070823]

In [77]:
# Napierの数表の項番を計算する
list(map(napier5, _))

[1234.0, 5677.999999999997]

In [78]:
# 対数法則から積を求める
napier5(sum(__))*10**5

7006651.999999997

In [85]:
# 検算
1234*5678

7006652

:::{seealso}
* [Compound interest - Wikipedia](https://en.wikipedia.org/wiki/Compound_interest)
    - (1655&mdash;1705) [Jacob Bernoulli - Wikipedia](https://en.wikipedia.org/wiki/Jacob_Bernoulli)
:::

### 出版された表を再現する

Napierは、$R=7$の場合の数表を作成した。

In [20]:
!pip install -U astropy

  Using cached astropy-6.1.1-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (10 kB)
  Using cached pyerfa-2.0.1.4-cp39-abi3-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (5.7 kB)
Using cached astropy-6.1.1-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (10.2 MB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.9/1.9 MB 36.8 MB/s eta 0:00:0000:01
Using cached pyerfa-2.0.1.4-cp39-abi3-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (738 kB)


In [21]:
from astropy.coordinates import Angle
from astropy import units

In [22]:
Angle('{:d}°{:d}′{:d}″'.format(1,2,3)).radian

0.018049613347708025

In [23]:
import math
import numpy as np
#from scoping import scoping
import pandas as pd

In [24]:
def napier(dms_d=0, r=7):
    for dms_m in range(60):
        a = Angle('{:d}°{:d}′{:d}″'.format(dms_d,dms_m,0))
        yield (a,
               "{:8.0f}".format(math.sin(a.radian)*10**r),
               "{:8.0f}".format(math.log(math.sin(a.radian), 1-1/(10**r)) if a.radian else float('inf')),
               "{:8.0f}".format(math.log(math.tan(a.radian), 1-1/(10**r)) if a.radian else float('inf')),
               "{:8.0f}".format(math.log(math.cos(a.radian), 1-1/(10**r))),
               "{:8.0f}".format(math.cos(a.radian)*10**r),
               Angle('{:d}°{:d}′{:d}″'.format(90,0,0))-Angle('{:d}°{:d}′{:d}″'.format(dms_d,dms_m,0)))

```{note}
角度が$0\degree$から$1\degree$の正弦、余弦とそれらの対数、および正接の対数の値。角度は一分刻み
```

In [25]:
pd.DataFrame(napier(dms_d=0), columns=['d','sin','log(sin(d))','log(tan(d))','log(cos(d))', 'cos(d)', '90-d'])

,d,sin,log(sin(d)),log(tan(d)),log(cos(d)),cos(d),90-d
0,0d00m00s,0,inf,inf,-0,10000000,90d00m00s
1,0d01m00s,2909,81425711,81425711,0,10000000,89d59m00s
2,0d02m00s,5818,74494240,74494239,2,9999998,89d58m00s
3,0d03m00s,8727,70439590,70439586,4,9999996,89d57m00s
4,0d04m00s,11636,67562771,67562764,7,9999993,89d56m00s
5,0d05m00s,14544,65331336,65331326,11,9999989,89d55m00s
6,0d06m00s,17453,63508123,63508107,15,9999985,89d54m00s
7,0d07m00s,20362,61966618,61966597,21,9999979,89d53m00s
8,0d08m00s,23271,60631306,60631279,27,9999973,89d52m00s
9,0d09m00s,26180,59453478,59453444,34,9999966,89d51m00s


```{note}
角度が$89\degree$から$90\degree$の正弦、余弦とそれらの対数、および正接の対数の値。角度は一分刻み
```

In [26]:
pd.DataFrame(napier(dms_d=89), columns=['d','sin','log(sin(d))','log(tan(d))','log(cos(d))', 'cos(d)', '90-d'])

,d,sin,log(sin(d)),log(tan(d)),log(cos(d)),cos(d),90-d
0,89d00m00s,9998477,1523,-40481252,40482775,174524,1d00m00s
1,89d01m00s,9998527,1473,-40649357,40650830,171616,0d59m00s
2,89d02m00s,9998577,1423,-40820334,40821758,168707,0d58m00s
3,89d03m00s,9998625,1375,-40994284,40995659,165799,0d57m00s
4,89d04m00s,9998673,1327,-41171312,41172639,162890,0d56m00s
5,89d05m00s,9998720,1280,-41351528,41352808,159982,0d55m00s
6,89d06m00s,9998766,1234,-41535050,41536284,157073,0d54m00s
7,89d07m00s,9998812,1188,-41722002,41723190,154165,0d53m00s
8,89d08m00s,9998856,1144,-41912513,41913657,151256,0d52m00s
9,89d09m00s,9998900,1100,-42106723,42107824,148348,0d51m00s
